README
NAME
    AnyEvent::TFTPd - Trivial File Transfer Protocol daemon

VERSION
    0.1303

DESCRIPTION
    I suddenly decided to leave any AnyEvent code (including
    AnyEvent::Handle::UDP), due to a community and development model that is
    indeed very hard to work with. If you want this module, please drop me
    mail and I'll hand over the maintenance.

    Update: AnyEvent::Handle states:

        # too many clueless people try to use udp and similar sockets
        # with AnyEvent::Handle, do them a favour.

    So instead of hacking the source of AnyEvent I'm simply abandoning this
    ship.

    This module handles TFTP request in an AnyEvent environment. It will set
    up a socket, handled by AnyEvent::Handle::UDP. Every time a new packet
    has arrived, it will call "on_read()" which handles the request. The
    rest is up to the AnyEvent::TFTPd::Connection object, which is possible
    to customize either by subclassing or modifying with Moose.

    Want timeout mechanism? See AnyEvent::TFTPd::CheckConnections.

SYNOPSIS
     package My::AnyEvent::Connection;
     use Moose;
     extends 'AnyEvent::TFTPd::Connection';

     sub _build_filehandle {
        my $self = shift;
        my $file = $self->file;

        # ...

        return $filehandle;
     }

     package main;

     my $tftpd = AnyEvent::TFTPd->new(
                     address => 'localhost',
                     port => 69,
                     connection_class => 'My::AnyEvent::Connection',
                     max_connections => 100,
                 )->setup or die $@;

ATTRIBUTES
  address
    Holds the address this server should bind to. Default is "127.0.0.1".

  port
    Holds the default port this server should listen to. Default is 69.

  retries
    This value will never be changes. It is used as default for the
    AnyEvent::TFTPd::Connection::retries attribute.

    Default number of retries are 3. (default value is subject for change)

  connection_class
    This string holds the classname where the connection objects should be
    constructed from. The default AnyEvent::TFTPd::Connection class is quite
    useless without subclassing it. See "SYNOPSIS" for more details.

  max_connections
    The max concurrent connections this object can handle. Used inside
    "on_connect()" to decide if a new connection should be establised or
    not.

    Setting this to zero (the default) means that the server should handle
    unlimited connections.

  _connections
     $connection_obj = $self->get_connection(peername);
     $connection_obj = $self->add_connection($connection_obj);
     @connections = $self->get_all_connections;

    This attribute holds a hash-ref, where the keys are "peername()" of the
    connections, and the values point to AnyEvent::TFTPd::Connection
    objects. Use the delegated methods listed above to access this
    attribute.

  _handle
     $io_socket_inet = $self->socket;
     $packed = $self->peername;

    This attribute holds an instance of AnyEvent::Handle::UDP, which handles
    the methods listed above.

METHODS
  setup
    This method will prepare the handle/socket for incoming connections. It
    will return c<$self> on success and 0 on failure. Check $@ for a full
    error message on failure.

    Return value $self allows you to chain "new()" and "setup()".

  on_read
    This hook is called each time data is received from a peer host. It will
    parse the datagram received and act accordingly.

  on_connect
    This method returns a new AnyEvent::TFTPd::Connection object for a new
    connection. This method is called when either a RRQ/WRQ opcode is
    received in "on_read()".

    This method might skip these steps if no more connections are available.
    This is computed by comparing the number of connections and
    "max_connections".

  on_error
    This hook is called from the handler, when something unexpected has
    happened. See AnyEvent::Handle for details.

COPYRIGHT & LICENSE
    Copyright 2007 Jan Henning Thorsen, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Jan Henning Thorsen "jhthorsen at cpan.org"



Changes
Revision history for AnyEvent-TFTPd

0.1303   Wed Apr 20 12:12:50 CEST 2011
       * Trying again to skip all tests...

0.1302   Sun Feb  6 18:38:10 CET 2011
       * Skipping all tests, because of failing POD tests
         http://www.cpantesters.org/cpan/report/ca83a9bc-2e09-11e0-a839-9e11f1b0ebf2

0.1301   Sun Jun 13 18:32:21 CEST 2010
       * Code got sabotaged

0.13     Fri May 21 00:09:50 CEST 2010
       * Decided to leave AnyEvent

0.12     Wed May 19 11:57:20 CEST 2010
       * Fix Makefile.PL: Missing dependecies
       * More builders
       * Add lazy_build to attributes

0.11     Sun May 16 18:26:03 CEST 2010
       * Add support for retry
         - add retries attribute to AnyEvent::TFTPd
         - add retries attribute to AnyEvent::TFTPd::Connection
         - on_read() checks number of retries
         - various actions will decrease retries
       * Add CheckConnections role for AnyEvent::TFTPd
         - will decrease retry on timeout
         - will remove connection when exhausted
       * Add last_seen_peer attribute to AnyEvent::TFTPd::Connection
       * Remove constants - replaced some with builders
       * Fix typos in pod

0.10     Sun May 16 12:40:18 CEST 2010
       * Add read/write
       * Add error handling both ways
       * Add max_connections
       * Add optional custom connection_class
       * Add connection cleanup on send/receive complete
       * TODO: clean up timed out connections



lib/AnyEvent/TFTPd/Connection.pm
package AnyEvent::TFTPd::Connection;

=head1 NAME

AnyEvent::TFTPd::Connection - Represents one connection to TFTPd

=head1 DESCRIPTION

=head1 SYNOPSIS

=cut

use Moose;
use Socket;

use constant MIN_BLKSIZE => 512;
use constant OPCODE_RRQ => 1;
use constant OPCODE_WRQ => 2;
use constant OPCODE_DATA => 3;
use constant OPCODE_ACK => 4;
use constant OPCODE_ERROR => 5;
use constant OPCODE_OACK => 6;

use overload (
    q("") => sub { $_[0]->peername },
    fallback => 1,
);

our %ERROR_CODES = (
    not_defined         => [0, 'Not defined, see error message'],
    unknown_opcode      => [0, 'Unknown opcode: %s'],
    no_connection       => [0, 'No connection'],
    file_not_found      => [1, 'File not found'],
    access_violation    => [2, 'Access violation'],
    disk_full           => [3, 'Disk full or allocation exceeded'],
    illegal_operation   => [4, 'Illegal TFTP operation'],
    unknown_transfer_id => [5, 'Unknown transfer ID'],
    file_exists         => [6, 'File already exists'],
    no_such_user        => [7, 'No such user'],
);

=head1 ATTRIBUTES

=head2 peername

Holds the sockaddr of the remote host.

=cut

has peername => (
    is => 'ro',
    isa => 'Str',
    required => 1,
);

=head2 address

Holds a human readable version of the address part of L</peername>.

=cut

has address => (
    is => 'ro',
    isa => 'Str',
    init_arg => undef,
    lazy => 1,
    default => sub { inet_ntoa +(sockaddr_in $_[0]->peername )[1] },
);

=head2 port

Holds a human readable version of the port part of L</peername>.

=cut

has port => (
    is => 'ro',
    isa => 'Int',
    init_arg => undef,
    lazy => 1,
    default => sub { (sockaddr_in $_[0]->peername )[0] },
);

=head2 opcode

A numeric representation of the opcode which initiated the connection.

=cut

has opcode => (
    is => 'ro',
    isa => 'Int',
    required => 1,
);

=head2 mode

Either "ascii" or "octet" or empty string if unknown.

=cut

has mode => (
    is => 'ro',
    isa => 'Str',
    default => '',
);

=head2 file

A string representing the requested file from client.

=cut

has file => (
    is => 'ro',
    isa => 'Str',
    default => '',
);

=head2 filehandle

The filehandle used to read/write data from/to client.

=cut

has filehandle => (
    is => 'rw',
    isa => 'Maybe[GlobRef]',
    lazy_build => 1,
);

sub _build_filehandle {
    my $self = shift;
    my $mode = $self->opcode eq OPCODE_WRQ ? '>' : '<';
    my $file = $self->file;
    my $FH;

    if($mode eq '<') {
        if(!-r $file) {
            $self->logf(warn => 'File cannot be read: %s', $file);
            return;
        }
    }
    else {
        if(-e $file) {
            $self->logf(warn => 'File exists: %s', $file);
            return;
        }
    }

    unless(open $FH, $mode, $file) {
        $self->logf(error => 'Read %s: %s', $file, $!);
        return;
    }

    return $FH;
}

=head2 rfc

Contains extra parameters the client has provided. These parameters are
stored in a hash ref.

=cut

has rfc => (
    is => 'ro',
    isa => 'ArrayRef',
    default => sub { [] },
);

=head2 server

A reference back to the L<AnyEvent::TFTPd> object.

=cut

has server => (
    is => 'ro',
    isa => 'Object',
    required => 1,
    handles => [qw/ socket /],
);

=head2 connected_at

The time the connection was established: Epoch timestamp.

=cut

has connected_at => (
    is => 'ro',
    isa => 'Int',
    default => sub { time },
);

=head2 packet_number

The current packet number received/sent.

=cut

has packet_number => (
    is => 'ro',
    isa => 'Int',
    traits => ['Counter'],
    default => 1,
    handles => {
        inc_packet_number => 'inc',
    },
);

=head2 last_seen_peer

The last time data was received from the peer: Epoch timestamp.
This value is used to calculate when to time out a connection.

=cut

has last_seen_peer => (
    is => 'rw',
    isa => 'Int',
    default => -1,
    trigger => sub { $_[0]->retries($_[0]->server->retries) },
);

=head2 retries

The number of retries left before aborting the transmission.
This number will be reset to L<AnyEvent::TFTPd::retries> each time
L</last_seen_peer> is updated.

=cut

has retries => (
    is => 'rw',
    isa => 'Int',
    traits => ['Counter'],
    lazy => 1,
    default => sub { $_[0]->server->retries },
    handles => {
        dec_retries => 'dec',
    },
);

=head1 METHODS

=head2 send_packet

This method will send a packet of data from L</filehandle>
to client, identified by L</peername>. The packet is calculated
using the C<MIN_BLKSIZE> and L</packet_number>. Returns 1 on
success, 2 if this is the last packet to be sent, 0 if something
went wrong and -1 if no more data is available from filehandle.

=cut

sub send_packet {
    my $self = shift;
    my $FH = $self->filehandle;
    my $n = $self->packet_number;
    my $data;

    if(not $FH) {
        $self->send_error('file_not_found');
        return 0;
    }
    if(not seek $FH, ($n - 1) * MIN_BLKSIZE, 0) {
        $self->logf(error => 'Seek %s: %s', $self->file, $!);
        $self->send_error('file_not_found');
        return 0;
    }
    if(not defined read $FH, $data, MIN_BLKSIZE) {
        $self->logf(error => 'Read %s: %s', $self->file, $!);
        $self->send_error('file_not_found');
        return 0;
    }
    if(0 == length $data and 1 < $self->packet_number) {
        $self->logf(debug => 'Peer has successfully received %s', $self->file);
        return -1;
    }

    $self->socket->send(
        pack('nna*', OPCODE_DATA, $self->packet_number, $data),
        MSG_DONTWAIT,
        $self->peername,
    ) or do {
        $self->dec_retries;
        $self->logf(error => 'Send %s: %s', $self->file, $!);
        return 0;
    };

    #$self->logf(trace => 'Sent packet n=%i', $self->packet_number);

    return length $data < MIN_BLKSIZE ? 2 : 1;
}

=head2 receive_ack

This method will receive a datagram and unwraps the packet number from
it using C<unpack("n")>. It will increase the L</packet_number> if
the received packet number matches L</packet_number>.

Will always call L</send_packet> and return the value it returns. A
return value of -1 indicates that the last ACK was received and the
connection can be "shut down".

=cut

sub receive_ack {
    my $self = shift;
    my($n) = unpack 'n', shift;

    if($n == $self->packet_number) {
        #$self->logf(trace => 'Received ack n=%i', $n);
        $self->inc_packet_number;
        $self->last_seen_peer(time);
    }
    else {
        $self->logf(warn => 'Wrong packet number: %i != %i', $n, $self->packet_number);
        $self->dec_retries;
    }

    return $self->send_packet;
}

=head2 receive_packet

This method will receive a datagram and unwraps the packet number and
body from it using C<unpack("na*")>. It stores the data in the current
filehandle if the packet number equals L</packet_number>. It returns
1 on success, 0 on failure and -1 if this was the last packet to be
received. The latter indicates that it is safe for this connection to
"shut down".

=cut

sub receive_packet {
    my $self = shift;
    my($n, $data) = unpack 'na*', shift;
    my $FH = $self->filehandle;

    unless($FH) {
        $self->send_error('illegal_operation');
        return 0;
    }
    unless($n == $self->packet_number) {
        $self->logf(warn => 'Wrong packet number: %i != %i', $n, $self->packet_number);
        $self->dec_retries;
        return $self->send_ack;
    }

    print(
        $FH $data
    ) or do {
        $self->logf(error => 'Write %s: %s', $self->file, $!);
        $self->send_error('illegal_operation');
        return 0;
    };

    $self->last_seen_peer(time);
    $self->inc_packet_number;
    #$self->logf(trace => 'Received packet n=%i', $n);

    if(length $data < MIN_BLKSIZE) {
        $self->logf(debug => 'Peer has successfully sent %s', $self->file);
        $self->send_ack;
        return -1;
    }

    return $self->send_ack;
}

=head2 send_ack

This method is called inside L</receive_packet()> and is used to tell
the peer that a packet is received.

=cut

sub send_ack {
    my $self = shift;
    my $n = $self->packet_number - 1;

    $self->socket->send(
        pack('nna*', OPCODE_ACK, $n, ''),
        MSG_DONTWAIT,
        $self->peername,
    ) or do {
        $self->logf(error => 'Send %s: %s', $self->file, $!);
        $self->dec_retries;
        return 0;
    };

    #$self->logf(trace => 'Sent ack n=%i', $n);

    return 1;
}

=head2 send_error

Takes a "name" indicating a type of error, which is looked up from the
C<%ERROR_CODES> variable (see the source for details). Falls back to
"not_defined", if an invalid name was passed on. Packs the data from
C<%ERROR_CODES> and pass it to the remote client. Returns 1 on success
and 1 on failure.

=cut

sub send_error {
    my $self = shift;
    my $name = shift || q();
    my $ERROR = $ERROR_CODES{$name} || $ERROR_CODES{'not_defined'};

    $self->socket->send(
        pack('nnZ*', OPCODE_ERROR, @$ERROR),
        MSG_DONTWAIT,
        $self->peername,
    ) or do {
        $self->logf(error => 'Send error=%s: %s', $name, $!);
        $self->dec_retries;
        return 0;
    };

    return 1;
}

=head2 logf

 $self->logf($level => @message);

Receives a C<$level> and a list of strings, suitable for C<printf()>:
The first element in C<@message> is a format and the rest (if exists) will
be the replacement for C<%s> and friends in the format. The level has the
same name as defined for L<Log::Log4perl>.

=cut

sub logf {
    my $self = shift;
    my $level = shift;
    my $format = shift;

    if($AnyEvent::TFTPd::DEBUG) {
        printf STDERR "%s %s:%s $level> $format\n",
            time,
            $self->address,
            $self->port,
            @_,
    }

    return 1;
}

=head1 BUGS

=head1 COPYRIGHT & LICENSE

=head1 AUTHOR

See L<Top::Module>.

=cut

1;


lib/AnyEvent/TFTPd/CheckConnections.pm
package AnyEvent::TFTPd::CheckConnections;

=head1 NAME

AnyEvent::TFTPd::CheckConnections - Role for AnyEvent::TFTPd for timeout checking

=head1 DESCRIPTION

This L<Moose> role can be applied to L<AnyEvent::TFTPd> which again will
provide an L<AnyEvent> timer to check for timed out connections.
See also L<MooseX::Traits> for ways to construct this object with this
role applied.

=head1 SYNOPSIS

 use AnyEvent::TFTPd;
 use AnyEvent::TFTPd::CheckConnections;

 my $tftpd = AnyEvent::TFTPd->new(...);

 # apply to all instances of AnyEvent::TFTPd
 AnyEvent::TFTPd::CheckConnections->meta->apply(AnyEvent::TFTPd->meta);

 # apply only two this instance
 AnyEvent::TFTPd::CheckConnections->meta->apply($tftpd);

 $tftpd->setup(timeout => 10); # unless set in constructor

=cut

use AnyEvent;
use Moose::Role;

#requires qw/ setup logf /;

=head1 ATTRIBUTES

=head2 timeout

Holds the timeout set in either constructor or when calling L</setup()>.

=cut

has timeout => (
    is => 'ro',
    isa => 'Int',
    writer => '_set_timeout',
    lazy_build => 1,
);

sub _build_timeout { 3 }

has _check_connection_timer => (
    is => 'ro',
    isa => 'Object',
    lazy_build => 1,
);

sub _build__check_connection_timer {
    my $self = shift;

    AnyEvent->timer(
        after => $self->timeout,
        interval => $self->timeout,
        cb => sub { $self->check_connections },
    );
}

=head1 METHODS

=head2 after setup

This method modifier will start the timed event which calls
L</check_connections()>.

=cut

after setup => sub {
    my $self = shift;
    my %args = @_;

    if(exists $args{'timeout'}) {
        $self->_set_timeout($args{'timeout'});
    }

    $self->_check_connection_timer;
};

=head2 check_connections

Will loop through all connections to see if any has timed out. If so,
decrease the number of retries and retry sending the data to peer if any
retries are possible. If not, remove the connection.

It is very important that the L<AnyEvent::TFTPd> object does not have
too many connections, since it will cause this loop to stall the program.
When that is said, the number of connections will probably cause problems
to the rest of the program, before slowing down this method.

=cut

sub check_connections {
    my $self = shift;
    my $time = time - $self->timeout;
    my $n = 0;

    for my $connection ($self->get_all_connections) {
        my $timeout = 0;

        if($connection->last_seen_peer <= $time) {
            $timeout = 1;
            $connection->dec_retries;
        }

        if($connection->retries < 0) {
            $connection->logf(error => 'Retries has exceeded');
            $self->delete_connection($connection);
            $n++;
        }
        elsif($timeout) {
            $connection->logf(error => 'Timeout has exceeded');

            if($connection->opcode == &AnyEvent::TFTPd::OPCODE_RRQ) {
                $connection->send_packet;
            }
            elsif($connection->opcode == &AnyEvent::TFTPd::OPCODE_WRQ) {
                $connection->send_ack;
            }
        }
    }

    return $n;
}

=head1 BUGS

=head1 COPYRIGHT & LICENSE

=head1 AUTHOR

See L<AnyEvent::TFTPd>.

=cut

1;


lib/AnyEvent/TFTPd.pm
package AnyEvent::TFTPd;

=head1 NAME

AnyEvent::TFTPd - Trivial File Transfer Protocol daemon

=head1 VERSION

0.1303

=head1 DESCRIPTION

I suddenly decided to leave any L<AnyEvent> code (including
L<AnyEvent::Handle::UDP>), due to a community and development model
that is indeed very hard to work with. If you want this module, please
drop me mail and I'll hand over the maintenance.

Update: L<AnyEvent::Handle> states:

    # too many clueless people try to use udp and similar sockets
    # with AnyEvent::Handle, do them a favour.

So instead of hacking the source of AnyEvent I'm simply abandoning this
ship.

This module handles TFTP request in an L<AnyEvent> environment. It
will set up a socket, handled by L<AnyEvent::Handle::UDP>. Every time
a new packet has arrived, it will call L</on_read()> which handles
the request. The rest is up to the L<AnyEvent::TFTPd::Connection>
object, which is possible to customize either by subclassing or
modifying with L<Moose>.

Want timeout mechanism? See L<AnyEvent::TFTPd::CheckConnections>.

=head1 SYNOPSIS

 package My::AnyEvent::Connection;
 use Moose;
 extends 'AnyEvent::TFTPd::Connection';

 sub _build_filehandle {
    my $self = shift;
    my $file = $self->file;

    # ...

    return $filehandle;
 }

 package main;

 my $tftpd = AnyEvent::TFTPd->new(
                 address => 'localhost',
                 port => 69,
                 connection_class => 'My::AnyEvent::Connection',
                 max_connections => 100,
             )->setup or die $@;

=cut

use Moose;
use AnyEvent::Handle::UDP;
use AnyEvent::TFTPd::Connection;
use IO::Socket::INET;

use constant OPCODE_RRQ => 1;
use constant OPCODE_WRQ => 2;
use constant OPCODE_DATA => 3;
use constant OPCODE_ACK => 4;
use constant OPCODE_ERROR => 5;
use constant OPCODE_OACK => 6;

our $VERSION = eval '0.1303';
our $DEBUG = 0;

=head1 ATTRIBUTES

=head2 address

Holds the address this server should bind to. Default is "127.0.0.1".

=cut

has address => (
    is => 'ro',
    isa => 'Str',
    lazy_build => 1,
);

sub _build_address { '127.0.0.1' }

=head2 port

Holds the default port this server should listen to. Default is 69.

=cut

has port => (
    is => 'ro',
    isa => 'Str',
    lazy_build => 1,
);

sub _build_port { 69 }

=head2 retries

This value will never be changes. It is used as default for the
L<AnyEvent::TFTPd::Connection::retries> attribute.

Default number of retries are 3. (default value is subject for change)

=cut

has retries => (
    is => 'ro',
    isa => 'Int',
    lazy_build => 1,
);

sub _build_retries { 3 }

=head2 connection_class

This string holds the classname where the connection objects should
be constructed from. The default L<AnyEvent::TFTPd::Connection> class
is quite useless without subclassing it. See L</SYNOPSIS> for more details.

=cut

has connection_class => (
    is => 'ro',
    isa => 'Str',
    default => 'AnyEvent::TFTPd::Connection',
);

=head2 max_connections

The max concurrent connections this object can handle. Used inside
L</on_connect()> to decide if a new connection should be establised or
not.

Setting this to zero (the default) means that the server should handle
unlimited connections.

=cut

has max_connections => (
    is => 'ro',
    isa => 'Int',
    lazy_build => 1,
);

sub _build_max_connections { 0 }

=head2 _connections

 $connection_obj = $self->get_connection(peername);
 $connection_obj = $self->add_connection($connection_obj);
 @connections = $self->get_all_connections;

This attribute holds a hash-ref, where the keys are C<peername()> of the
connections, and the values point to L<AnyEvent::TFTPd::Connection> objects.
Use the delegated methods listed above to access this attribute.

=cut

has _connections => (
    is => 'ro',
    isa => 'HashRef',
    traits => ['Hash'],
    default => sub { {} },
    handles => {
        get_connection => 'get',
        get_all_connections => 'values',
        delete_connection => 'delete',
    },
);

__PACKAGE__->meta->add_method(_clear_connections => sub {
    %{ $_[0]->_connections } = ();
});
__PACKAGE__->meta->add_method(add_connection => sub {
    return $_[0]->_connections->{"$_[1]"} = $_[1];
});

=head2 _handle

 $io_socket_inet = $self->socket;
 $packed = $self->peername;

This attribute holds an instance of L<AnyEvent::Handle::UDP>, which
handles the methods listed above.

=cut

has _handle => (
    is => 'ro',
    isa => 'AnyEvent::Handle::UDP',
    lazy_build => 1,
    handles => [qw/ socket peername /],
);

sub _build__handle {
    my $self = shift;
    
    return AnyEvent::Handle::UDP->new(
        listen => join(':', $self->address, $self->port),
        read_size => 1428, # or 512?
        on_error => sub { $self->on_error(@_) },
        on_read => sub { $self->on_read(@_) },
    );
}

=head1 METHODS

=head2 setup

This method will prepare the handle/socket for incoming connections.
It will return c<$self> on success and 0 on failure. Check C<$@> for
a full error message on failure.

Return value C<$self> allows you to chain C<new()> and C<setup()>.

=cut

sub setup {
    my $self = shift;
    my %args = @_; # used in CheckConnections

    eval {
        $self->_handle;
    } or do {
        return 0;
    };

    return $self;
}

=head2 on_read

This hook is called each time data is received from a peer host.
It will parse the datagram received and act accordingly.

=cut

sub on_read {
    my $self = shift;
    my $handle = shift;
    my $datagram = $handle->rbuf;
    my $opcode = unpack 'n', substr $datagram, 0, 2, '';
    my $connection;

    $handle->{'rbuf'} = ''; # clear buffer

    # ===============
    # Init connection

    if($opcode == OPCODE_RRQ) {
        if($connection = $self->on_connect($opcode, $datagram)) {
            $connection->send_packet;
            return 1;
        }
        return 0;
    }
    elsif($opcode == OPCODE_WRQ) {
        if($connection = $self->on_connect($opcode, $datagram)) {
            $connection->send_ack;
            return 1;
        }
        return 0;
    }

    # ======================
    # Connection in progress

    $connection = $self->get_connection($handle->peername);

    if(!$connection) {
        $connection = $self->connection_class->new(
                          opcode => $opcode,
                          peername => $handle->peername,
                          server => $self,
                      );

        $connection->logf(warn => 'Connection is not established');
        $connection->send_error('no_connection');

        return 0;
    }

    if($opcode == OPCODE_ACK) {
        if($connection->receive_ack($datagram) == -1) {
            $self->delete_connection($connection);
            return -1;
        }
    }
    elsif($opcode == OPCODE_DATA) {
        if($connection->receive_packet($datagram) == -1) {
            $self->delete_connection($connection);
            return -1;
        }
    }
    elsif($opcode == OPCODE_ERROR) {
        my($code, $msg) = unpack 'nZ*', $datagram;
        $connection->logf(error => 'Error from client: %s/%s', $code, $msg);
    }
    else {
        $connection->logf(error => 'Unknown opcode: %i', $opcode);
        $connection->send_error('unknown_opcode');
        return 0;
    }

    if($connection->retries == -1) {
        $connection->logf(error => 'Retries has exceeded');
        $self->delete_connection($connection);
        return 0;
    }

    return 1;
}

=head2 on_connect

This method returns a new L<AnyEvent::TFTPd::Connection> object for
a new connection. This method is called when either a RRQ/WRQ opcode
is received in L</on_read()>.

This method might skip these steps if no more connections are
available. This is computed by comparing the number of connections
and L</max_connections>.

=cut

sub on_connect {
    my $self = shift;
    my $opcode = shift;
    my $datagram = shift;
    my $max_connections = $self->max_connections;
    my($file, $mode, @rfc) = split "\0", $datagram;
    my $connection;

    $connection = $self->connection_class->new(
                      opcode => $opcode,
                      peername => $self->peername,
                      server => $self,
                      file => $file,
                      mode => lc($mode),
                      rfc => \@rfc,
                  );

    if(0 < $max_connections and $max_connections <= $self->get_all_connections) {
        $connection->logf(debug => 'Max connection limit reached');
        return;
    }
    else {
        $connection->logf(debug => 'Connection established');
        return $self->add_connection($connection);
    }
}

=head2 on_error

This hook is called from the handler, when something unexpected has
happened. See L<AnyEvent::Handle> for details.

=cut

sub on_error {
    my $self = shift;
    my $handle = shift;
    my $fatal = shift;
    my $message = shift;

    warn $message if $DEBUG;

    if($fatal) {
        # rebuild handle
        $self->_clear_connections;
        $self->_clear_handle;
        $self->_handle;
    }

    return 1;
}

=head1 COPYRIGHT & LICENSE

Copyright 2007 Jan Henning Thorsen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR

Jan Henning Thorsen C<< jhthorsen at cpan.org >>

=cut

1;
